CodeReview of 18 from 4

Fisher4537 opened the pull request:
Implemented Barcode Scanner with [MLKit](https://firebase.google.com/docs/ml-kit/android/read-barcodes) library.
Hello Porro, this is a copy of the pull request that you can find [here](https://github.com/PrandiniUniPD/OCRCamera/pull/34). For the code review it is y not necessary that you review all the changes: it would be too much. You can ignore all the .xml file changes and review ONLY the .java files. Let me know if you have better proposal.

Fisher4537 committed:
added BarcodeExtractor to read barcode, bug on thread managment: critical lock held

Fisher4537 committed:
added a copy of missing xml from drawable-v24 to drawable and vice versa

Fisher4537 committed:
added the files mentioned in the last commit

Fisher4537 committed:
Merge branch 'CorrectResourcesBug' into BarcodeReader

Fisher4537 committed:
added constructor of BarcodeExtractor class

Fisher4537 committed:
bugfix barcodeView setted under OCRTextView

Fisher4537 committed:
edited landscaper layout of ResultActivity, added 'Barcode:' beforethe barcode number in ResultActivity

Fisher4537 committed:
added comments to simplify code reading

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java:
```diff
@@ -33,6 +44,8 @@
      */
     private TextView mOCRTextView;
 
+    private TextView mBarcodeView;
```
Maybe here you can write what's mBarcodeView is like it's done in line 43 for mOCRTextView, and in line 50 for last photo.

Fisher4537 replied:
What do you think about "barcodeTextView"? It seems better to me!

thomasporro replied:
I was thinking to simply add a comment above your variable to make it more readable

Fisher4537 replied:
I've changed the name in "barcodeTextView" and added a comment, to the variable to make it clearer!

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java:
```diff
@@ -56,16 +73,18 @@ public void onClick(View view) {
         SharedPreferences prefs = getSharedPreferences("prefs", MODE_PRIVATE);
         String pathImage = prefs.getString("imagePath", null);
         String OCRText = prefs.getString("text", null);
+        String barcodeText = prefs.getString("barcodeText",null);
```
Maybe you can modify line 76 to add the extracting the last barcode read in the comment

Fisher4537 replied:
Which comment are you talking about? Do you mean that barcodeText should be the string of the last barcode readed?

thomasporro replied:
The comment in line 76. Just like this:
`\\Get the image path, text and barcode of the last image from preferences `

Fisher4537 replied:
Oh nice! Now I can see what you mean!

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java:
```diff
@@ -180,5 +213,79 @@ protected void onPreExecute() {
+        @Override
+        protected String doInBackground(Bitmap... bitmaps){
+            String rawBarcode = "";
+            //BarcodeExtractor barcodeExtractor = ;
```
It's just my opinion, but here maybe you can specify why the method is commented. So it's easier to understand if you want to modify the code in future or just you deleted the line

Fisher4537 replied:
Thanks, I've delated the comment! It has been my forgetfulness

thomasporro replied:
The comment is still there

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java:
```diff
@@ -180,5 +213,79 @@ protected void onPreExecute() {
+        protected String progressMessage;
+
+        /**
+         * Costructor initialize:
```
Maybe you can specify which variables the constructor initialize.

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java:
```diff
@@ -180,5 +213,79 @@ protected void onPreExecute() {
+    private class BarcodeReader extends AsyncTask<Bitmap, Void, String>{
+
+        protected ProgressDialog progressDialog; //TODO: progressDialog is deprecated, replace with ProgressBar
+        protected TextView barcodeTextView; //
```
You probably forgot to delete the "//"

Fisher4537 replied:
Sure! Thanks!

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java:
```diff
@@ -180,5 +213,79 @@ protected void onPreExecute() {
+            if(lastPhoto != null) {
+                rawBarcode = new BarcodeExtractor().getTextFromImg(lastPhoto); //see BarcodeExtractor class
+
+                if(rawBarcode.equals("")) //No barcode found case
```
In my opinion the brace should be put in the same line of the "if". Here a my suggestion
```suggestion
                //No barcode found case
                if(rawBarcode.equals("")){
```

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java:
```diff
@@ -180,5 +213,79 @@ protected void onPreExecute() {
+                            barcodeTextView.setText(finalTextRecognized);
+                        }
+                    });
+                }
```
The same of the above comment. I leave down below a suggestion:
```suggestion
                } else {    // found at least one barcode case
```

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/TextExtractor.java:
```diff
@@ -25,7 +25,9 @@
      * Constructor
      */
-    TextExtractor() {    }
+    TextExtractor() {
```
It's just a my preference, but i prefer the constructor and the braces in the same line

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/TextExtractor.java:
```diff
@@ -54,23 +56,24 @@ private String extractText(Bitmap img)
+                        //analogous to signal
+                        latch.countDown();
+                    }
+                });
```
The same question I've asked in BarcodeExtractor.java

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java:
```diff
@@ -0,0 +1,119 @@
+
+/**
+ * Class which try to extract the content of a barcode in an image
+ * @author Luca Perali (inspired to Pietro Prandini's TextExtractor)
```
Nice referral

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java:
```diff
@@ -0,0 +1,119 @@
+     * (The method has the same name of OCRInterface methods for eventual future implementation of a common
+     * interface)
+     *
+     * @param img Bitmap image with or without barcode in it
```
Maybe you can modify the specs, because "image with or without barcode in it", is just the image we want to analize. The fact that if there is a barcode is already reported in @return

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java:
```diff
@@ -0,0 +1,119 @@
+        Log.d(TAG, "extractTextFromImage");
+        long initialTime = java.lang.System.currentTimeMillis();
+
+        // Firebase implementation tutorial at https://firebase.google.com/docs/ml-kit/android/read-barcodes
```
The links to the documentations are very nice

Fisher4537 replied:
Thanks! I think that it's very usefull to have a pointer on the same documentation!

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java:
```diff
@@ -0,0 +1,119 @@
+                        latch.countDown();
+                    }
+                });
+        /*         .addOnFailureListener(new OnFailureListener() {
```
It's just my opinion, but here maybe you can specify why the method is commented. So it's easier to understand if you want to modify the code in future or just you deleted the line

Fisher4537 replied:
You are right! I'm Fixing it! Thanks

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java:
```diff
@@ -0,0 +1,119 @@
+                        //analogous to signal
+                        latch.countDown();
+                    }
+                });
```
This is a question: the brace reffered to .addOnSuccessListener or Task<List<FirebaseVisionBarcode>>?

Fisher4537 replied:
![image](https://user-images.githubusercontent.com/39212279/49365918-4da05980-f6e8-11e8-95c7-d7dd2aa0d59f.png)
Well, as you see in the image the yellow braces contain the explicit parameters of addOnSuccessListener method

thomasporro replied:
👍🏼

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java:
```diff
@@ -0,0 +1,119 @@
+
+        // if barcode is not detected yet
+        if(!task.isSuccessful()) {
+            try
```
You should use the syntax you've already used in other parts of the code. Like this:

```suggestion
           try{ 
                //the things you want to do
           } else {
                //other things
           }
```

thomasporro replied:
And the same is for the catch statement

Fisher4537 replied:
Bad abit! Thanks for the correction. Added for the next commit!

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java:
```diff
@@ -0,0 +1,119 @@
+     * @author Luca Perali
+     */
+    private static String getInfoFromBarcode(List<FirebaseVisionBarcode> barcodes) {
+        //StringBuilder result = new StringBuilder();
```
It's just my opinion, but here maybe you can specify why these methods are commented. So it's easier to understand if you want to modify the code in future, you used for debugging or just you deleted the lines

Fisher4537 replied:
I agree! I forgot also this one commented line! Thanks

thomasporro replied:
I would prefer if the commented methods had an explanation, for example if you made them for debugging

thomasporro replied:
This comment  ( `//StringBuilder result = new StringBuilder();` ) and the following in my opinion still a problem in the code until you don't explain their function or you delete them

Fisher4537 replied:
Pushing the remotion of the comments!

thomasporro reviewed (commented):
Sorry if I took so long to do the review but in these days I was very busy.
Btw the code is ok, I've commented some parts where the syntax is different from the code in general, and some parts that contains methods-comments.
But in general the code is very good.

Fisher4537 committed:
fixed some identation problems, deleted useless comment

Fisher4537 commented:
Thank to you, I've fixed the document style bug and deleted useless comments. I'm going to commit the last comment modification and then if you agree we could close the pull request!

Fisher4537 committed:
comment modification

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/BarcodeExtractor.java:
```diff
@@ -0,0 +1,118 @@
+package unipd.se18.ocrcamera;
+
+
```
They reported in my review that according to our style document:

>Imports are ordered as follows:
>All static imports in a single block.
>All non-static imports in a single block.
>If there are both static and non-static imports, a single blank line separates the two blocks. There are no other blank lines between import statements.

thomasporro commented on app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java:
```diff
@@ -33,6 +44,8 @@
      */
     private TextView mOCRTextView;
 
```
I was thinking to add an "explanation" like this:
```suggestion
     /**
      * The TextView of the extracted barcode from the captured photo.
      */
```

Fisher4537 replied:
Thanks for the suggestion! I'm fixing it!

thomasporro reviewed (commented):
The code seems better, there are only a few things to change and in my opinion it's ok

Fisher4537 committed:
removed comment in line 247 of ResultActivity and added comment to explain barcodeTextView meaning

Fisher4537 committed:
removed hard code

thomasporro reviewed (commented):
The changes you made were fine! There is only a few commented methods to fix

Fisher4537 committed:
removed useless comments

Fisher4537 commented:
Thanks for the review Thomas, I appreciated your tenacity! Closing the pull request
